bndtools.core/src/bndtools/central/QueryWhiteboard.java

```diff
@@ -0,0 +1,54 @@
+package bndtools.central;
+
+import org.eclipse.search.ui.IQueryListener;
+import org.eclipse.search.ui.NewSearchUI;
+import org.eclipse.ui.IWorkbench;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+/**
+ * This component adds whiteboard facility to IQueryListeners.
+ */
+@Component(immediate = true)
+public class QueryWhiteboard implements ServiceTrackerCustomizer<IQueryListener, IQueryListener> {
+	@Reference
+	IWorkbench										wb;
+
+	ServiceTracker<IQueryListener, IQueryListener>	listeners;
```

Comment:
This was one of the things that I hoped you would comment on. :smile:

I originally tried bind/unbind/updated, but I found that I was getting NPEs. I realised that this was because the get called before the reference to IWorkbench was satisifed and before the component was activated. It got complicated to track them as they were removed/added before activation, and then to track them differently after activation (when the workspace is available), and as I thought about how to handle this problem I realised that this was exactly the sort of problem that `ServiceTracker` was designed to solve.

However, I admit to not being a DS expert. If there is an easy way to configure DS so that the dynamic reference updates are queued by DS and only delivered *after* activation, then that would work. (If there is no such feature in DS, maybe we could consider it a feature request?)

If you have a better suggestion, I'd love to hear it.

Reply:
Implemented @rotty3000 's suggestion for both the resource and query whiteboards and removed the `ServiceTracker`s.